In [26]:
!pip3 install torch
!pip3 install numpy

In [27]:
import torch
import numpy as np

In [28]:
x = torch.arange(6).reshape(3,2)
x

tensor([[0, 1],
        [2, 3],
        [4, 5]])

In [29]:
x[1,1]

tensor(3)

In [30]:
x[:,1]

tensor([1, 3, 5])

In [31]:
x[:,1:]

tensor([[1],
        [3],
        [5]])

In [32]:
x = torch.arange(10)
x 

tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [33]:
x.view(2,5)

tensor([[0, 1, 2, 3, 4],
        [5, 6, 7, 8, 9]])

In [34]:
x.reshape(2,5)

tensor([[0, 1, 2, 3, 4],
        [5, 6, 7, 8, 9]])

In [35]:
x = torch.arange(10)
x

tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [36]:
x.shape

torch.Size([10])

In [37]:
z = x.view(2,5)
z

tensor([[0, 1, 2, 3, 4],
        [5, 6, 7, 8, 9]])

In [38]:
a = torch.Tensor([1,2,3])
b = torch.Tensor([4,5,6])


In [39]:
a + b
a

tensor([1., 2., 3.])

In [40]:
b

tensor([4., 5., 6.])

In [41]:
torch.add(a,b)

tensor([5., 7., 9.])

In [42]:
a.mul(b) # -> a + b

tensor([ 4., 10., 18.])

In [43]:
a.mul_(b)  # -> a = a + b

tensor([ 4., 10., 18.])

# Part 2

In [44]:
# Create two tensors
tensor1 = torch.tensor([1, 2, 3])
tensor2 = torch.tensor([4, 5, 6])

# Compute the dot product
dot_product = tensor1.dot(tensor2)

# 1*4 + 2*5 + 3*6 = 4 + 10 + 18 = 32
print(dot_product)

tensor(32)


In [45]:
# Create two matrices
matrix1 = torch.tensor([[1, 2], [3, 4]])
matrix2 = torch.tensor([[5, 6], [7, 8]])

# Perform matrix multiplication
result = torch.mm(matrix1, matrix2)

#  [[1*5 + 2*7, 1*6 + 2*8],
#  [3*5 + 4*7, 3*6 + 4*8]]
#  = [[19, 22],
#    [43, 50]]

print(result)


tensor([[19, 22],
        [43, 50]])


In [46]:
matrix1 @ matrix2 # same like torch.mm

tensor([[19, 22],
        [43, 50]])

In PyTorch, the torch.norm() function is used to compute various types of vector norms or matrix norms. It calculates the magnitude or size of a tensor according to a specified norm type.

The torch.norm() function can be called in different ways depending on the desired norm calculation. Here are a few common usages:

In [47]:
# Compute L2 norm of a vector
vector = torch.Tensor([1, 2, 3])
l2_norm = torch.norm(vector, p=2)
print(l2_norm)

tensor(3.7417)


In this example, the L2 norm (Euclidean norm) of the vector [1, 2, 3] is calculated using torch.norm() with p=2. The result is approximately 3.7417.

In [48]:
# Compute Frobenius norm of a matrix
matrix = torch.Tensor([[1, 2], [3, 4]])
frobenius_norm = torch.norm(matrix, p='fro')

print(frobenius_norm)


tensor(5.4772)


In this example, the element-wise L1 norm of the tensor [-1, -2, 3, 4] is calculated using torch.norm() with p=1 and dim=0. The result is 10, which represents the sum of the absolute values of the tensor elements.

The p parameter in torch.norm() specifies the type of norm to compute. It can take different values such as:

```code
p=1 for L1 norm
p=2 for L2 norm (default)
'fro' for Frobenius norm (matrix norm)
float('inf') for infinity norm (maximum absolute value)
```
Additionally, the dim parameter can be specified for calculating norms along a specific dimension of a tensor. If dim is not provided, the norm is computed over the entire tensor.

In [49]:
import torch

# Create a tensor
tensor = torch.tensor([[1, 2, 3], [4, 5, 6]])

# Get the number of elements
num_elements = torch.numel(tensor)

print(num_elements)


6


In this example, the tensor has 6 elements in total (2 rows and 3 columns), and torch.numel() returns the value 6.

You can also use tensor.numel() to achieve the same result:

In [50]:
import torch

# Create a tensor
tensor = torch.tensor([[1, 2, 3], [4, 5, 6]])

# Get the number of elements
num_elements = tensor.numel()

print(num_elements)

6
